diff --git a/Source/CMakeVersion.cmake b/Source/CMakeVersion.cmake
index f25a4fd57d..e222a0998e 100644
--- a/Source/CMakeVersion.cmake
+++ b/Source/CMakeVersion.cmake
@@ -80,5 +80,9 @@ else()
   set(CMake_VERSION_SUFFIX "")
 endif()
 if(CMake_VERSION_IS_DIRTY)
-  string(APPEND CMake_VERSION "-dirty")
+  if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
+    string(APPEND CMake_VERSION "-zos")
+  else()
+    string(APPEND CMake_VERSION "-dirty")
+  endif()
 endif()
