diff --git a/Source/cmGeneratorTarget.cxx b/Source/cmGeneratorTarget.cxx
index 7d1fcf3e5c..16f74c0d06 100644
--- a/Source/cmGeneratorTarget.cxx
+++ b/Source/cmGeneratorTarget.cxx
@@ -396,7 +396,8 @@ const char* cmGeneratorTarget::GetOutputTargetType(
         switch (artifact) {
           case cmStateEnums::RuntimeBinaryArtifact:
             // A DLL shared library is treated as a runtime target.
-            return "RUNTIME";
+            // On z/OS it's treated as library target.
+            return this->Makefile->IsOn("OS390") ? "LIBRARY" : "RUNTIME";
           case cmStateEnums::ImportLibraryArtifact:
             // A DLL import library is treated as an archive target.
             return "ARCHIVE";
