diff --git a/Modules/CMakeDetermineCompilerABI.cmake b/Modules/CMakeDetermineCompilerABI.cmake
index 13bfeeced7..8b0ba5680e 100644
--- a/Modules/CMakeDetermineCompilerABI.cmake
+++ b/Modules/CMakeDetermineCompilerABI.cmake
@@ -39,6 +39,16 @@ function(CMAKE_DETERMINE_COMPILER_ABI lang src)
       # from which we might detect implicit link libraries.
       list(APPEND CMAKE_FLAGS "-DCMAKE_${lang}_STANDARD_LIBRARIES=")
     endif()
+
+    # Add Flag for z/OS compilers for ascii encoding of string literals.
+    if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+      if(CMAKE_${lang}_COMPILER_ID MATCHES "Clang")
+        string(APPEND COMPILE_DEFINITIONS " -fzos-le-char-mode=ascii")
+      elseif(CMAKE_${lang}_COMPILER_ID MATCHES "zOS")
+        string(APPEND COMPILE_DEFINITIONS " -qascii")
+      endif()
+    endif()
+
     __TestCompiler_setTryCompileTargetType()
 
     # Avoid failing ABI detection on warnings.
