diff --git a/Modules/CMakeDetermineCompilerABI.cmake b/Modules/CMakeDetermineCompilerABI.cmake
index 230eec8b..ef377b27 100644
--- a/Modules/CMakeDetermineCompilerABI.cmake
+++ b/Modules/CMakeDetermineCompilerABI.cmake
@@ -71,6 +71,21 @@ function(CMAKE_DETERMINE_COMPILER_ABI lang src)
     if(DEFINED CMAKE_${lang}_VERBOSE_LINK_FLAG)
       list(APPEND LINK_OPTIONS "${CMAKE_${lang}_VERBOSE_LINK_FLAG}")
     endif()
+    # Add Flag for z/OS compilers for ascii encoding of string literals.
+    if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+      if(CMAKE_${lang}_COMPILER_ID MATCHES "Clang")
+        get_filename_component(COMPILER_BASENAME "${CMAKE_${lang}_COMPILER}" NAME)
+        if(COMPILER_BASENAME STREQUAL "ibm-clang" OR COMPILER_BASENAME STREQUAL "ibm-clang++")
+          string(APPEND COMPILE_DEFINITIONS " -m32 -fzos-le-char-mode=ascii")
+          list(APPEND LINK_OPTIONS " -m32")
+        else()
+          string(APPEND COMPILE_DEFINITIONS " -fzos-le-char-mode=ascii")
+        endif()
+      elseif(CMAKE_${lang}_COMPILER_ID MATCHES "zOS")
+        string(APPEND COMPILE_DEFINITIONS " -qascii")
+        list(APPEND LINK_OPTIONS " -qascii")
+      endif()
+    endif()
     if(lang MATCHES "^(CUDA|HIP)$")
       if(CMAKE_CUDA_ARCHITECTURES STREQUAL "native")
         # We are about to detect the native architectures, so we do
