diff --git a/Utilities/cmlibuv/src/unix/core.c b/Utilities/cmlibuv/src/unix/core.c
index 0f997a2d..db663915 100644
--- a/Utilities/cmlibuv/src/unix/core.c
+++ b/Utilities/cmlibuv/src/unix/core.c
@@ -63,7 +63,7 @@
 # include <crt_externs.h>
 # include <mach-o/dyld.h> /* _NSGetExecutablePath */
 # define environ (*_NSGetEnviron())
-#else /* defined(__APPLE__) && !TARGET_OS_IPHONE */
+#elif !defined(__MVS__)
 extern char** environ;
 #endif /* !(defined(__APPLE__) && !TARGET_OS_IPHONE) */
 
@@ -124,9 +124,6 @@ STATIC_ASSERT(offsetof(uv_buf_t, len) == offsetof(struct iovec, iov_len));
 
 /* https://github.com/libuv/libuv/issues/1674 */
 int uv_clock_gettime(uv_clock_id clock_id, uv_timespec64_t* ts) {
-#if defined(__APPLE__) && MAC_OS_X_VERSION_MIN_REQUIRED < 101200
-  return UV_ENOSYS;
-#else
   struct timespec t;
   int r;
 
@@ -151,7 +148,6 @@ int uv_clock_gettime(uv_clock_id clock_id, uv_timespec64_t* ts) {
   ts->tv_nsec = t.tv_nsec;
 
   return 0;
-#endif
 }
 
 
@@ -209,9 +205,9 @@ void uv_close(uv_handle_t* handle, uv_close_cb close_cb) {
 
   case UV_FS_EVENT:
     uv__fs_event_close((uv_fs_event_t*)handle);
-#if defined(__sun) || defined(__MVS__)
+#if defined(__sun)
     /*
-     * On Solaris, illumos, and z/OS we will not be able to dissociate the
+     * On Solaris, illumos, we will not be able to dissociate the
      * watcher for an event which is pending delivery, so we cannot always call
      * uv__make_close_pending() straight away. The backend will call the
      * function once the event has cleared.
@@ -602,6 +598,10 @@ int uv__accept(int sockfd) {
  * by making the system call directly. Musl libc is unaffected.
  */
 int uv__close_nocancel(int fd) {
+#if defined(__MVS__)
+  // See https://github.ibm.com/open-z/node/pull/410.
+  SAVE_ERRNO(epoll_file_close(fd));
+#endif
 #if defined(__APPLE__)
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wdollar-in-identifier-extension"
@@ -648,9 +648,6 @@ int uv__close_nocheckstdio(int fd) {
 
 int uv__close(int fd) {
   assert(fd > STDERR_FILENO);  /* Catch stdio close bugs. */
-#if defined(__MVS__)
-  SAVE_ERRNO(epoll_file_close(fd));
-#endif
   return uv__close_nocheckstdio(fd);
 }
 
@@ -1103,7 +1100,9 @@ static int uv__getrusage(int who, uv_rusage_t* rusage) {
 
   if (getrusage(who, &usage))
     return UV__ERR(errno);
-
+#if defined(__MVS__)
+  memset(rusage, 0, sizeof(*rusage));
+#endif
   rusage->ru_utime.tv_sec = usage.ru_utime.tv_sec;
   rusage->ru_utime.tv_usec = usage.ru_utime.tv_usec;
 
@@ -1179,6 +1178,9 @@ int uv_getrusage_thread(uv_rusage_t* rusage) {
   return uv__getrusage(RUSAGE_LWP, rusage);
 #elif defined(RUSAGE_THREAD)
   return uv__getrusage(RUSAGE_THREAD, rusage);
+#elif defined(__MVS__)
+  // TODO(gabylb): zos - https://github.ibm.com/open-z/node-build/issues/1217
+  return UV_ENOTSUP;
 #endif  /* defined(__APPLE__) */
   return UV_ENOTSUP;
 }
@@ -1497,9 +1499,7 @@ int uv_os_get_group(uv_group_t* grp, uv_uid_t gid) {
   /* Copy the groupname */
   grp->groupname = gr_mem;
   memcpy(grp->groupname, gp.gr_name, name_size);
-#ifndef __clang_analyzer__ /* deadcode.DeadStores */
   gr_mem += name_size;
-#endif
 
   /* Copy the gid */
   grp->gid = gp.gr_gid;
@@ -1616,6 +1616,10 @@ int uv_os_setenv(const char* name, const char* value) {
   if (setenv(name, value, 1) != 0)
     return UV__ERR(errno);
 
+#if defined(__MVS__)
+  __update_envar_settings(name);
+#endif
+
   return 0;
 }
 
@@ -1627,6 +1631,10 @@ int uv_os_unsetenv(const char* name) {
   if (unsetenv(name) != 0)
     return UV__ERR(errno);
 
+#if defined(__MVS__)
+  __update_envar_settings(name);
+#endif
+
   return 0;
 }
 
@@ -1723,7 +1731,6 @@ int uv_os_setpriority(uv_pid_t pid, int priority) {
 #endif
 }
 
-#ifndef CMAKE_BOOTSTRAP
 /**
  * If the function succeeds, the return value is 0.
  * If the function fails, the return value is non-zero.
@@ -1731,6 +1738,11 @@ int uv_os_setpriority(uv_pid_t pid, int priority) {
  * So the output parameter priority is actually the nice value.
 */
 int uv_thread_getpriority(uv_thread_t tid, int* priority) {
+#ifdef __MVS__
+  // TODO(gabylb): zos - no pthread_getschedparam() on z/OS.
+  *priority = 0;
+  return 0;
+#else
   int r;
   int policy;
   struct sched_param param;
@@ -1758,8 +1770,8 @@ int uv_thread_getpriority(uv_thread_t tid, int* priority) {
 
   *priority = param.sched_priority;
   return 0;
-}
 #endif
+}
 
 #ifdef __linux__
 static int set_nice_for_calling_thread(int priority) {
@@ -1778,13 +1790,12 @@ static int set_nice_for_calling_thread(int priority) {
 }
 #endif
 
-#ifndef CMAKE_BOOTSTRAP
 /**
  * If the function succeeds, the return value is 0.
  * If the function fails, the return value is non-zero.
 */
 int uv_thread_setpriority(uv_thread_t tid, int priority) {
-#if !defined(__GNU__)
+#if !defined(__GNU__) && !defined(__MVS__)
   int r;
   int min;
   int max;
@@ -1855,7 +1866,6 @@ int uv_thread_setpriority(uv_thread_t tid, int priority) {
   return 0;
 #endif  /* !defined(__GNU__) */
 }
-#endif
 
 int uv_os_uname(uv_utsname_t* buffer) {
   struct utsname buf;
@@ -2015,8 +2025,10 @@ int uv__search_path(const char* prog, char* buf, size_t* buflen) {
   while (token != NULL) {
     snprintf(trypath, sizeof(trypath) - 1, "%s/%s", token, prog);
     if (realpath(trypath, abspath) == abspath) {
-      /* Check the match is executable */
-      if (access(abspath, X_OK) == 0) {
+      struct stat info;
+      /* Check the match is not a directory, and is executable. */
+      if (stat(abspath, &info) == 0 && !S_ISDIR(info.st_mode) &&
+          access(abspath, X_OK) == 0) {
         abspath_size = strlen(abspath);
 
         *buflen -= 1;
@@ -2074,10 +2086,6 @@ unsigned int uv_available_parallelism(void) {
     rc = uv__cpu_count(&set);
 #elif defined(__MVS__)
   rc = __get_num_online_cpus();
-  if (rc < 1)
-    rc = 1;
-
-  return (unsigned) rc;
 #elif defined(__FreeBSD__)
   cpuset_t set;
 
@@ -2131,8 +2139,10 @@ unsigned int uv_available_parallelism(void) {
   }
 #endif /* __linux__ */
 
+#if !defined(__MVS__)
   if (rc < 0)
     rc = sysconf(_SC_NPROCESSORS_ONLN);
+#endif /* !__MVS__ */
 
 #ifdef __linux__
   {
@@ -2192,3 +2202,27 @@ int uv__sock_reuseport(int fd) {
 
   return 0;
 }
+
+/* Check if rlimit has any expressed restrictions on usable memory. */
+uint64_t uv__get_rlimit_max_memory(void) {
+  struct rlimit rl;
+  uint64_t result = 0;
+  uint64_t rlimit_value;
+
+#if defined(RLIMIT_AS)
+  if (getrlimit(RLIMIT_AS, &rl) == 0 && rl.rlim_cur != RLIM_INFINITY) {
+    rlimit_value = rl.rlim_cur;
+    result = rlimit_value;
+  }
+#endif
+
+#if defined(RLIMIT_DATA)
+  if (getrlimit(RLIMIT_DATA, &rl) == 0 && rl.rlim_cur != RLIM_INFINITY) {
+    rlimit_value = rl.rlim_cur;
+    if (result == 0 || rlimit_value < result)
+      result = rlimit_value;
+  }
+#endif
+
+  return result;
+}
