diff --git a/Modules/CMakeDetermineCCompiler.cmake b/Modules/CMakeDetermineCCompiler.cmake
index 8beebc5e57..92bddede2d 100644
--- a/Modules/CMakeDetermineCCompiler.cmake
+++ b/Modules/CMakeDetermineCCompiler.cmake
@@ -93,7 +93,13 @@ else()
     )
 endif()
 if(CMAKE_C_COMPILER_TARGET)
-  set(CMAKE_C_COMPILER_ID_TEST_FLAGS_FIRST "-c --target=${CMAKE_C_COMPILER_TARGET}")
+  if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+    # z/OS object files sometimes don't have continuous string literals, therefore
+    # leads to some compiler info not being correctly parsed, so we should not use -c.
+    set(CMAKE_C_COMPILER_ID_TEST_FLAGS_FIRST "--target=${CMAKE_C_COMPILER_TARGET}")
+  else()
+    set(CMAKE_C_COMPILER_ID_TEST_FLAGS_FIRST "-c --target=${CMAKE_C_COMPILER_TARGET}")
+  endif()
 endif()
 # Build a small source file to identify the compiler.
 if(NOT CMAKE_C_COMPILER_ID_RUN)
@@ -178,7 +184,8 @@ if (NOT _CMAKE_TOOLCHAIN_PREFIX)
 
     # if "llvm-" is part of the prefix, remove it, since llvm doesn't have its own binutils
     # but uses the regular ar, objcopy, etc. (instead of llvm-objcopy etc.)
-    if ("${_CMAKE_TOOLCHAIN_PREFIX}" MATCHES "(.+-)?llvm-$")
+    # remove "ibm-" prefix for the same reason above.
+    if ("${_CMAKE_TOOLCHAIN_PREFIX}" MATCHES "(.+-)?(llvm|ibm)-$")
       set(_CMAKE_TOOLCHAIN_PREFIX ${CMAKE_MATCH_1})
     endif ()
   elseif(CMAKE_C_COMPILER_ID MATCHES "TI")
diff --git a/Modules/CMakeDetermineCXXCompiler.cmake b/Modules/CMakeDetermineCXXCompiler.cmake
index 40934ecac5..61b4050a10 100644
--- a/Modules/CMakeDetermineCXXCompiler.cmake
+++ b/Modules/CMakeDetermineCXXCompiler.cmake
@@ -95,7 +95,13 @@ else()
 endif()
 
 if(CMAKE_CXX_COMPILER_TARGET)
-  set(CMAKE_CXX_COMPILER_ID_TEST_FLAGS_FIRST "-c --target=${CMAKE_CXX_COMPILER_TARGET}")
+  if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+    # z/OS object files sometimes don't have continuous string literals, therefore
+    # leads to some compiler info not being correctly parsed, so we should not use -c.
+    set(CMAKE_C_COMPILER_ID_TEST_FLAGS_FIRST "--target=${CMAKE_C_COMPILER_TARGET}")
+  else()
+    set(CMAKE_C_COMPILER_ID_TEST_FLAGS_FIRST "-c --target=${CMAKE_C_COMPILER_TARGET}")
+  endif()
 endif()
 
 # Build a small source file to identify the compiler.
@@ -183,7 +189,8 @@ if (NOT _CMAKE_TOOLCHAIN_PREFIX)
 
     # if "llvm-" is part of the prefix, remove it, since llvm doesn't have its own binutils
     # but uses the regular ar, objcopy, etc. (instead of llvm-objcopy etc.)
-    if ("${_CMAKE_TOOLCHAIN_PREFIX}" MATCHES "(.+-)?llvm-$")
+    # remove "ibm-" prefix for the same reason above.
+    if ("${_CMAKE_TOOLCHAIN_PREFIX}" MATCHES "(.+-)?(llvm|ibm)-$")
       set(_CMAKE_TOOLCHAIN_PREFIX ${CMAKE_MATCH_1})
     endif ()
   elseif(CMAKE_CXX_COMPILER_ID MATCHES "TI")
diff --git a/Modules/CMakeDetermineCompilerABI.cmake b/Modules/CMakeDetermineCompilerABI.cmake
index 13bfeeced7..8b0ba5680e 100644
--- a/Modules/CMakeDetermineCompilerABI.cmake
+++ b/Modules/CMakeDetermineCompilerABI.cmake
@@ -39,6 +39,16 @@ function(CMAKE_DETERMINE_COMPILER_ABI lang src)
       # from which we might detect implicit link libraries.
       list(APPEND CMAKE_FLAGS "-DCMAKE_${lang}_STANDARD_LIBRARIES=")
     endif()
+
+    # Add Flag for z/OS compilers for ascii encoding of string literals.
+    if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+      if(CMAKE_${lang}_COMPILER_ID MATCHES "Clang")
+        string(APPEND COMPILE_DEFINITIONS " -fzos-le-char-mode=ascii")
+      elseif(CMAKE_${lang}_COMPILER_ID MATCHES "zOS")
+        string(APPEND COMPILE_DEFINITIONS " -qascii")
+      endif()
+    endif()
+
     __TestCompiler_setTryCompileTargetType()
 
     # Avoid failing ABI detection on warnings.
diff --git a/Modules/CMakeDetermineSystem.cmake b/Modules/CMakeDetermineSystem.cmake
index 386be73f68..b50b4a52c8 100644
--- a/Modules/CMakeDetermineSystem.cmake
+++ b/Modules/CMakeDetermineSystem.cmake
@@ -29,7 +29,7 @@
 # cygwin                        CYGWIN_NT-5.1
 # MSYS                          MSYS_NT-6.1
 # MacOSX                        Darwin
-
+# z/OS                          OS/390
 
 # find out on which system cmake runs
 if(CMAKE_HOST_UNIX)
@@ -53,7 +53,7 @@ if(CMAKE_HOST_UNIX)
         OUTPUT_STRIP_TRAILING_WHITESPACE
         ERROR_QUIET)
     endif()
-    if(CMAKE_HOST_SYSTEM_NAME MATCHES "Linux|CYGWIN.*|MSYS.*|^GNU$|Android")
+    if(CMAKE_HOST_SYSTEM_NAME MATCHES "Linux|CYGWIN.*|MSYS.*|^GNU$|Android|OS/390")
       execute_process(COMMAND ${CMAKE_UNAME} -m
         OUTPUT_VARIABLE CMAKE_HOST_SYSTEM_PROCESSOR
         RESULT_VARIABLE val
diff --git a/Modules/CMakeFindBinUtils.cmake b/Modules/CMakeFindBinUtils.cmake
index f778891b41..5e7fa860e8 100644
--- a/Modules/CMakeFindBinUtils.cmake
+++ b/Modules/CMakeFindBinUtils.cmake
@@ -173,7 +173,7 @@ else()
   set(_CMAKE_TAPI_NAMES "tapi")
 
   # Prepend toolchain-specific names.
-  if("${CMAKE_${_CMAKE_PROCESSING_LANGUAGE}_COMPILER_ID}" STREQUAL Clang)
+  if("${CMAKE_${_CMAKE_PROCESSING_LANGUAGE}_COMPILER_ID}" STREQUAL Clang AND NOT CMAKE_SYSTEM_NAME MATCHES "OS390")
     if("x${CMAKE_${_CMAKE_PROCESSING_LANGUAGE}_SIMULATE_ID}" STREQUAL "xMSVC")
       list(PREPEND _CMAKE_LINKER_NAMES "lld-link")
     elseif(NOT APPLE)
diff --git a/Modules/CMakePlatformId.h.in b/Modules/CMakePlatformId.h.in
index 32b71663ca..3289f50570 100644
--- a/Modules/CMakePlatformId.h.in
+++ b/Modules/CMakePlatformId.h.in
@@ -35,6 +35,9 @@
 #elif defined(_AIX) || defined(__AIX) || defined(__AIX__) || defined(__aix) || defined(__aix__)
 # define PLATFORM_ID "AIX"
 
+#elif defined(__MVS__)
+# define PLATFORM_ID "zOS"
+
 #elif defined(__hpux) || defined(__hpux__)
 # define PLATFORM_ID "HP-UX"
 
@@ -195,6 +198,9 @@
 # elif defined(__ICCSTM8__)
 #  define ARCHITECTURE_ID "STM8"
 
+# elif defined(__MVS__) && defined(__ARCH__)
+#  define ARCHITECTURE_ID "ARCH" STRINGIFY(__ARCH__)
+
 # else /* unknown architecture */
 #  define ARCHITECTURE_ID ""
 # endif
diff --git a/Modules/CheckTypeSize.cmake b/Modules/CheckTypeSize.cmake
index 01ce1d202a..bc7197de6e 100644
--- a/Modules/CheckTypeSize.cmake
+++ b/Modules/CheckTypeSize.cmake
@@ -143,6 +143,16 @@ function(__check_type_size_impl type var map builtin language)
   set(bin ${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/CheckTypeSize/${var}.bin)
   file(READ ${__check_type_size_dir}/CheckTypeSize.c.in src_content)
   string(CONFIGURE "${src_content}" src_content @ONLY)
+  # Add Flag for z/OS compilers for ascii encoding of string literals.
+  if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+    # Only need to modify CMAKE_C_FLAGS as CheckTypeSize.c.in is in C
+    set(CMAKE_C_FLAGS_SAVE ${CMAKE_C_FLAGS})
+    if(CMAKE_C_COMPILER_ID MATCHES "Clang")
+      string(APPEND CMAKE_C_FLAGS " -fzos-le-char-mode=ascii")
+    elseif(CMAKE_C_COMPILER_ID MATCHES "zOS")
+      string(APPEND CMAKE_C_FLAGS " -qASCII")
+    endif()
+  endif()
   try_compile(HAVE_${var} SOURCE_FROM_VAR "${src}" src_content
     COMPILE_DEFINITIONS ${CMAKE_REQUIRED_DEFINITIONS}
     LINK_OPTIONS ${CMAKE_REQUIRED_LINK_OPTIONS}
@@ -152,6 +162,10 @@ function(__check_type_size_impl type var map builtin language)
       "-DINCLUDE_DIRECTORIES:STRING=${CMAKE_REQUIRED_INCLUDES}"
     COPY_FILE ${bin}
     )
+  # Restore flags for z/OS compilers.
+  if(CMAKE_SYSTEM_NAME MATCHES "OS390")
+    set(CMAKE_C_FLAGS ${CMAKE_C_FLAGS_SAVE})
+  endif()
 
   if(HAVE_${var})
     # The check compiled.  Load information from the binary.
diff --git a/Modules/Compiler/Clang.cmake b/Modules/Compiler/Clang.cmake
index 46f5fc1941..c038ed1a20 100644
--- a/Modules/Compiler/Clang.cmake
+++ b/Modules/Compiler/Clang.cmake
@@ -52,8 +52,13 @@ else()
       endif()
     endif()
 
-    set(_CMAKE_${lang}_IPO_SUPPORTED_BY_CMAKE YES)
-    set(_CMAKE_${lang}_IPO_MAY_BE_SUPPORTED_BY_COMPILER YES)
+    if(ZOS)
+      set(_CMAKE_${lang}_IPO_SUPPORTED_BY_CMAKE NO)
+      set(_CMAKE_${lang}_IPO_MAY_BE_SUPPORTED_BY_COMPILER NO)
+    else()
+      set(_CMAKE_${lang}_IPO_SUPPORTED_BY_CMAKE YES)
+      set(_CMAKE_${lang}_IPO_MAY_BE_SUPPORTED_BY_COMPILER YES)
+    endif()
 
     string(COMPARE EQUAL "${CMAKE_${lang}_COMPILER_ID}" "AppleClang" __is_apple_clang)
     string(COMPARE EQUAL "${CMAKE_${lang}_COMPILER_ID}" "FujitsuClang" __is_fujitsu_clang)
diff --git a/Modules/Compiler/IBMClang-CXX.cmake b/Modules/Compiler/IBMClang-CXX.cmake
index 5431b17e20..16ba6e65f7 100644
--- a/Modules/Compiler/IBMClang-CXX.cmake
+++ b/Modules/Compiler/IBMClang-CXX.cmake
@@ -26,14 +26,19 @@ set(CMAKE_CXX14_STANDARD__HAS_FULL_SUPPORT ON)
 set(CMAKE_CXX14_STANDARD_COMPILE_OPTION "-std=c++14")
 set(CMAKE_CXX14_EXTENSION_COMPILE_OPTION "-std=gnu++14")
 
-if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 17.1.0)
+if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 17.1.0 OR
+    (CMAKE_SYSTEM_NAME STREQUAL "OS390" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 1.1.0))
   set(CMAKE_CXX17_STANDARD_COMPILE_OPTION  "-std=c++17")
   set(CMAKE_CXX17_EXTENSION_COMPILE_OPTION "-std=gnu++17")
   set(CMAKE_CXX20_STANDARD_COMPILE_OPTION  "-std=c++20")
   set(CMAKE_CXX20_EXTENSION_COMPILE_OPTION "-std=gnu++20")
 endif()
 
-__compiler_check_default_language_standard(CXX 17.1.0 17)
+if (CMAKE_SYSTEM_NAME STREQUAL "OS390")
+  __compiler_check_default_language_standard(CXX 1.1.0 14)
+else()
+  __compiler_check_default_language_standard(CXX 17.1.0 17)
+endif()
 
 set(CMAKE_CXX_COMPILE_OBJECT
   "<CMAKE_CXX_COMPILER> -x c++ <DEFINES> <INCLUDES> <FLAGS> -o <OBJECT> -c <SOURCE>")
diff --git a/Modules/Compiler/zOS-C.cmake b/Modules/Compiler/zOS-C.cmake
new file mode 100644
index 0000000000..73928346d2
--- /dev/null
+++ b/Modules/Compiler/zOS-C.cmake
@@ -0,0 +1,7 @@
+include(Compiler/zOS)
+__compiler_zos(C)
+set(CMAKE_C_FLAGS_RELEASE_INIT "${CMAKE_C_FLAGS_RELEASE_INIT} -DNDEBUG")
+set(CMAKE_C_FLAGS_MINSIZEREL_INIT "${CMAKE_C_FLAGS_MINSIZEREL_INIT} -DNDEBUG")
+
+# Raise severity of include file checks from warning to error for C.
+set(CMAKE_PLATFORM_CHECK_INCLUDE_C_FLAGS "-qseverity=e=ccn3296")
diff --git a/Modules/Compiler/zOS-CXX.cmake b/Modules/Compiler/zOS-CXX.cmake
new file mode 100644
index 0000000000..de82177f25
--- /dev/null
+++ b/Modules/Compiler/zOS-CXX.cmake
@@ -0,0 +1,39 @@
+include(Compiler/zOS)
+__compiler_zos(CXX)
+set(CMAKE_CXX_FLAGS_RELEASE_INIT "${CMAKE_CXX_FLAGS_RELEASE_INIT} -DNDEBUG")
+set(CMAKE_CXX_FLAGS_MINSIZEREL_INIT "${CMAKE_CXX_FLAGS_MINSIZEREL_INIT} -DNDEBUG")
+
+if(NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 2.3.1)
+  set(CMAKE_CXX98_STANDARD__HAS_FULL_SUPPORT ON)
+  set(CMAKE_CXX98_STANDARD_COMPILE_OPTION "-std=c++98")
+  set(CMAKE_CXX98_EXTENSION_COMPILE_OPTION "-std=gnu++98")
+
+  set(CMAKE_CXX11_STANDARD__HAS_FULL_SUPPORT ON)
+  set(CMAKE_CXX11_STANDARD_COMPILE_OPTION "-std=c++11")
+  set(CMAKE_CXX11_EXTENSION_COMPILE_OPTION "-std=gnu++11")
+
+  set(CMAKE_CXX14_STANDARD__HAS_FULL_SUPPORT ON)
+  set(CMAKE_CXX14_STANDARD_COMPILE_OPTION "-std=c++14")
+  set(CMAKE_CXX14_EXTENSION_COMPILE_OPTION "-std=gnu++14")
+endif()
+
+macro(cmake_record_cxx_compile_features)
+  list(APPEND CMAKE_CXX_COMPILE_FEATURES
+    cxx_std_98
+    cxx_std_11
+    cxx_std_14
+    )
+  _record_compiler_features(CXX "" CMAKE_CXX_COMPILE_FEATURES)
+endmacro()
+
+if(NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 2.3.1)
+  if (NOT CMAKE_CXX_COMPILER_FORCED)
+    if (NOT CMAKE_CXX_STANDARD_COMPUTED_DEFAULT)
+      message(FATAL_ERROR "CMAKE_CXX_STANDARD_COMPUTED_DEFAULT should be set for ${CMAKE_CXX_COMPILER_ID} (${CMAKE_CXX_COMPILER}) version ${CMAKE_CXX_COMPILER_VERSION}")
+    endif()
+    set(CMAKE_CXX_STANDARD_DEFAULT ${CMAKE_CXX_STANDARD_COMPUTED_DEFAULT})
+  elseif(NOT DEFINED CMAKE_CXX_STANDARD_DEFAULT)
+    # Compiler id was forced so just guess the default standard level.
+    set(CMAKE_CXX_STANDARD_DEFAULT 98)
+  endif()
+endif()
diff --git a/Modules/Compiler/zOS.cmake b/Modules/Compiler/zOS.cmake
new file mode 100644
index 0000000000..4a59c33f60
--- /dev/null
+++ b/Modules/Compiler/zOS.cmake
@@ -0,0 +1,33 @@
+
+#=============================================================================
+# Copyright 2002-2011 Kitware, Inc.
+#
+# Distributed under the OSI-approved BSD License (the "License");
+# see accompanying file Copyright.txt for details.
+#
+# This software is distributed WITHOUT ANY WARRANTY; without even the
+# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the License for more information.
+#=============================================================================
+# (To distribute this file outside of CMake, substitute the full
+#  License text for the above reference.)
+
+# This module is shared by multiple languages; use include blocker.
+if(__COMPILER_ZOS)
+  return()
+endif()
+set(__COMPILER_ZOS 1)
+
+macro(__compiler_zos lang)
+  # Feature flags.
+  set(CMAKE_${lang}_VERBOSE_FLAG "-V")
+
+  # Initial configuration flags.
+  set(CMAKE_${lang}_FLAGS_INIT "")
+  set(CMAKE_${lang}_FLAGS_DEBUG_INIT "-qdebug")
+  set(CMAKE_${lang}_FLAGS_MINSIZEREL_INIT "-O -DNDEBUG")
+  set(CMAKE_${lang}_FLAGS_RELEASE_INIT "-O3 -DNDEBUG")
+  set(CMAKE_${lang}_FLAGS_RELWITHDEBINFO_INIT "-O2 -qdebug -DNDEBUG")
+
+  set(CMAKE_DEPFILE_FLAGS_${lang} "-MF <DEPFILE> -qmakedep=gcc")
+endmacro()
diff --git a/Modules/FindThreads.cmake b/Modules/FindThreads.cmake
index 0fa6ae7661..1dbde75199 100644
--- a/Modules/FindThreads.cmake
+++ b/Modules/FindThreads.cmake
@@ -68,6 +68,10 @@ endif()
 
 # simple pthread test code
 set(PTHREAD_C_CXX_TEST_SOURCE [====[
+#ifdef __MVS__
+#define _UNIX03_THREADS
+#include <stddef.h>
+#endif
 #include <pthread.h>
 
 static void* test_func(void* data)
diff --git a/Modules/GenerateExportHeader.cmake b/Modules/GenerateExportHeader.cmake
index 9fdc783600..2b0f1b45b8 100644
--- a/Modules/GenerateExportHeader.cmake
+++ b/Modules/GenerateExportHeader.cmake
@@ -305,6 +305,8 @@ macro(_DO_SET_MACRO_VALUES TARGET_LIBRARY)
     if(WIN32 OR CYGWIN)
       set(DEFINE_EXPORT "__declspec(dllexport)")
       set(DEFINE_IMPORT "__declspec(dllimport)")
+    elseif(ZOS)
+      set(DEFINE_EXPORT "_Export")
     elseif(COMPILER_HAS_HIDDEN_VISIBILITY)
       set(DEFINE_EXPORT "__attribute__((visibility(\"default\")))")
       set(DEFINE_IMPORT "__attribute__((visibility(\"default\")))")
diff --git a/Modules/Platform/OS390.cmake b/Modules/Platform/OS390.cmake
new file mode 100644
index 0000000000..1773726a86
--- /dev/null
+++ b/Modules/Platform/OS390.cmake
@@ -0,0 +1,37 @@
+set(OS390 1)
+set(ZOS 1)
+
+set(CMAKE_STATIC_LIBRARY_PREFIX "lib")
+set(CMAKE_STATIC_LIBRARY_SUFFIX ".a")
+set(CMAKE_SHARED_LIBRARY_PREFIX "lib")
+set(CMAKE_SHARED_LIBRARY_SUFFIX ".so")
+set(CMAKE_LINK_LIBRARY_SUFFIX "")
+set(CMAKE_IMPORT_LIBRARY_PREFIX "lib")
+set(CMAKE_IMPORT_LIBRARY_SUFFIX ".x")
+set(CMAKE_EXECUTABLE_SUFFIX "")
+set(CMAKE_DL_LIBS "")
+
+set(CMAKE_FIND_LIBRARY_PREFIXES "lib")
+set(CMAKE_FIND_LIBRARY_SUFFIXES ".x" ".a")
+
+set(CMAKE_SHARED_LIBRARY_C_FLAGS "")
+
+if(CMAKE_C_COMPILER_ID MATCHES "Clang")
+  set(CMAKE_SHARED_LIBRARY_CREATE_C_FLAGS "-shared")
+elseif(CMAKE_C_COMPILER_ID MATCHES "zOS")
+  set(CMAKE_SHARED_LIBRARY_CREATE_C_FLAGS "-Wl,DLL")
+endif()
+
+set(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS "")
+
+# z/OS archiver (ar) cannot archive two file names with the same name
+# as part of the same archive command. We must make sure the
+# object file names are distinct.
+set(CMAKE_MANGLE_OBJECT_FILE_NAMES "ON")
+
+include(Platform/UnixPaths)
+
+# Allow options and operands to be interspersed for c89/cc/c++ utilities.
+set(_CC_CCMODE ON)
+set(_CXX_CCMODE ON)
+set(_C89_CCMODE ON)
diff --git a/README.rst b/README.rst
index 64e2353dd2..93f0946745 100644
--- a/README.rst
+++ b/README.rst
@@ -39,6 +39,7 @@ Supported Platforms
 * OpenBSD
 * Solaris
 * AIX
+* z/OS
 
 Other UNIX-like operating systems may work too out of the box, if not
 it should not be a major problem to port CMake to this platform.
diff --git a/Source/CMakeVersion.cmake b/Source/CMakeVersion.cmake
index f25a4fd57d..e222a0998e 100644
--- a/Source/CMakeVersion.cmake
+++ b/Source/CMakeVersion.cmake
@@ -80,5 +80,9 @@ else()
   set(CMake_VERSION_SUFFIX "")
 endif()
 if(CMake_VERSION_IS_DIRTY)
-  string(APPEND CMake_VERSION "-dirty")
+  if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
+    string(APPEND CMake_VERSION "-zos")
+  else()
+    string(APPEND CMake_VERSION "-dirty")
+  endif()
 endif()
diff --git a/Source/CursesDialog/form/form.h b/Source/CursesDialog/form/form.h
index b590c97852..7e36502d6c 100644
--- a/Source/CursesDialog/form/form.h
+++ b/Source/CursesDialog/form/form.h
@@ -50,7 +50,17 @@
 #    define _XOPEN_SOURCE_EXTENDED
 #   endif
 #  endif
+#  if defined(__MVS__)
+#   include <stdbool.h>
+#   define __bool_
+#  endif
+#  if defined(__MVS__) && defined(__cplusplus)
+extern "C" {
+#  endif
 #  include <curses.h>
+#  if defined(__MVS__) && defined(__cplusplus)
+}
+#  endif
 #  if defined(__hpux) && !defined(HAVE__XOPEN_SOURCE_EXTENDED)
 #   undef _XOPEN_SOURCE_EXTENDED
 #  endif
diff --git a/Source/CursesDialog/form/frm_driver.c b/Source/CursesDialog/form/frm_driver.c
index 9cbb12ff98..d8e1c3572b 100644
--- a/Source/CursesDialog/form/frm_driver.c
+++ b/Source/CursesDialog/form/frm_driver.c
@@ -865,7 +865,7 @@ static int Display_Or_Erase_Field(FIELD * field, bool bEraseFlag)
         Set_Field_Window_Attributes(field,win);
       else
         {
-#if defined(__LSB_VERSION__)
+#if defined(__LSB_VERSION__) || defined(__MVS__)
         /* getattrs() would be handy, but it is not part of LSB 4.0 */
         attr_t fwinAttrs;
         short  fwinPair;
diff --git a/Source/cmGeneratorTarget.cxx b/Source/cmGeneratorTarget.cxx
index 7d1fcf3e5c..16f74c0d06 100644
--- a/Source/cmGeneratorTarget.cxx
+++ b/Source/cmGeneratorTarget.cxx
@@ -396,7 +396,8 @@ const char* cmGeneratorTarget::GetOutputTargetType(
         switch (artifact) {
           case cmStateEnums::RuntimeBinaryArtifact:
             // A DLL shared library is treated as a runtime target.
-            return "RUNTIME";
+            // On z/OS it's treated as library target.
+            return this->Makefile->IsOn("OS390") ? "LIBRARY" : "RUNTIME";
           case cmStateEnums::ImportLibraryArtifact:
             // A DLL import library is treated as an archive target.
             return "ARCHIVE";
diff --git a/Source/cmInstallCommand.cxx b/Source/cmInstallCommand.cxx
index 7f5f15b3c9..b409fa62e5 100644
--- a/Source/cmInstallCommand.cxx
+++ b/Source/cmInstallCommand.cxx
@@ -134,6 +134,7 @@ public:
     const cmInstallCommandArguments* args) const;
   std::string GetManDestination(const cmInstallCommandArguments* args) const;
   std::string GetDocDestination(const cmInstallCommandArguments* args) const;
+  std::string GetImportLibrarySuffix(const cmInstallCommandArguments* args) const;
   std::string GetDestinationForType(const cmInstallCommandArguments* args,
                                     const std::string& type) const;
 
@@ -690,6 +691,9 @@ bool HandleTargetsMode(std::vector<std::string> const& args,
     return true;
   }
 
+  // Check whether this is a z/OS platform.
+  bool zos_platform = helper.GetImportLibrarySuffix(nullptr) == ".x";
+
   for (std::string const& tgt : targetList) {
 
     if (helper.Makefile->IsAlias(tgt)) {
@@ -834,19 +838,28 @@ bool HandleTargetsMode(std::vector<std::string> const& args,
               target, archiveArgs, true, helper.Makefile->GetBacktrace());
             artifactsSpecified = true;
           }
-          if (!runtimeArgs.GetDestination().empty()) {
+          if (!runtimeArgs.GetDestination().empty() && !zos_platform) {
             // The DLL uses the RUNTIME properties.
             runtimeGenerator = CreateInstallTargetGenerator(
               target, runtimeArgs, false, helper.Makefile->GetBacktrace());
             artifactsSpecified = true;
           }
-          if (!archiveGenerator && !runtimeGenerator) {
+          if (!libraryArgs.GetDestination().empty() && zos_platform) {
+            // The DLL uses the LIBRARY properties.
+            libraryGenerator = CreateInstallTargetGenerator(
+              target, libraryArgs, true, helper.Makefile->GetBacktrace());
+            artifactsSpecified = true;
+          }
+          if (!archiveGenerator && !runtimeGenerator && !libraryGenerator) {
             archiveGenerator = CreateInstallTargetGenerator(
               target, archiveArgs, true, helper.Makefile->GetBacktrace(),
               helper.GetArchiveDestination(nullptr));
             runtimeGenerator = CreateInstallTargetGenerator(
               target, runtimeArgs, false, helper.Makefile->GetBacktrace(),
               helper.GetRuntimeDestination(nullptr));
+            libraryGenerator = CreateInstallTargetGenerator(
+              target, libraryArgs, true, helper.Makefile->GetBacktrace(),
+              helper.GetLibraryDestination(nullptr));
           }
           if (runtimeDependencySet && runtimeGenerator) {
             runtimeDependencySet->AddLibrary(runtimeGenerator.get());
@@ -2461,6 +2474,12 @@ std::string Helper::GetDocDestination(
                               this->GetDataRootDestination(nullptr) + "/doc");
 }
 
+std::string Helper::GetImportLibrarySuffix(const cmInstallCommandArguments* args) const
+{
+  return this->Makefile->GetSafeDefinition("CMAKE_IMPORT_LIBRARY_SUFFIX");
+}
+
+
 std::string Helper::GetDestinationForType(
   const cmInstallCommandArguments* args, const std::string& type) const
 {
diff --git a/Source/cmLoadCommandCommand.cxx b/Source/cmLoadCommandCommand.cxx
index 1f18ff2eda..d529bc5c30 100644
--- a/Source/cmLoadCommandCommand.cxx
+++ b/Source/cmLoadCommandCommand.cxx
@@ -1,7 +1,7 @@
 /* Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
    file Copyright.txt or https://cmake.org/licensing for details.  */
 
-#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__)
+#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__) && !defined(__MVS__)
 // POSIX APIs are needed
 // NOLINTNEXTLINE(bugprone-reserved-identifier)
 #  define _POSIX_C_SOURCE 200809L
diff --git a/Source/cmMakefile.cxx b/Source/cmMakefile.cxx
index 0af0ed079a..d5d32914ce 100644
--- a/Source/cmMakefile.cxx
+++ b/Source/cmMakefile.cxx
@@ -4121,6 +4121,14 @@ int cmMakefile::ConfigureFile(const std::string& infile,
     // close the files before attempting to copy
     fin.close();
     fout.close();
+
+#if defined(__MVS__)
+    if ( !cmSystemTools::CopyFileAttributes(sinfile.c_str(),
+                                            tempOutputFile.c_str()) ) {
+      return 0;
+      }
+#endif
+
     if (!cmSystemTools::CopyFileIfDifferent(tempOutputFile, soutfile)) {
       this->IssueMessage(MessageType::FATAL_ERROR,
                          cmSystemTools::GetLastSystemError());
diff --git a/Source/cmStandardLexer.h b/Source/cmStandardLexer.h
index 27225280d6..77ba6c9067 100644
--- a/Source/cmStandardLexer.h
+++ b/Source/cmStandardLexer.h
@@ -8,7 +8,7 @@
 #  define _XOPEN_SOURCE 600
 #endif
 #if !defined(_POSIX_C_SOURCE) && !defined(_WIN32) && !defined(__sun) &&       \
-  !defined(__OpenBSD__)
+  !defined(__OpenBSD__) && !defined(__MVS__)
 /* POSIX APIs are needed */
 // NOLINTNEXTLINE(bugprone-reserved-identifier)
 #  define _POSIX_C_SOURCE 200809L
diff --git a/Source/cmTimestamp.cxx b/Source/cmTimestamp.cxx
index 7e47b4e51c..66f7bcb79d 100644
--- a/Source/cmTimestamp.cxx
+++ b/Source/cmTimestamp.cxx
@@ -1,7 +1,7 @@
 /* Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
    file Copyright.txt or https://cmake.org/licensing for details.  */
 
-#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__)
+#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__) && !defined(__MVS__)
 // POSIX APIs are needed
 // NOLINTNEXTLINE(bugprone-reserved-identifier)
 #  define _POSIX_C_SOURCE 200809L
diff --git a/Source/kwsys/SystemTools.cxx b/Source/kwsys/SystemTools.cxx
index 3bb78696c2..fbd22581e4 100644
--- a/Source/kwsys/SystemTools.cxx
+++ b/Source/kwsys/SystemTools.cxx
@@ -18,6 +18,10 @@
 #  define _WIN32_WINNT _WIN32_WINNT_VISTA
 #endif
 
+#if defined(__MVS__)
+#define _OPEN_SYS_FILE_EXT 1
+#endif
+
 #include "kwsysPrivate.h"
 #include KWSYS_HEADER(RegularExpression.hxx)
 #include KWSYS_HEADER(SystemTools.hxx)
@@ -85,7 +89,9 @@
 #  include <unistd.h>
 #  include <utime.h>
 #  ifndef __VMS
-#    include <sys/param.h>
+#    ifndef __MVS__
+#      include <sys/param.h>
+#    endif
 #    include <termios.h>
 #  endif
 #  include <csignal> /* sigprocmask */
@@ -2480,6 +2486,12 @@ SystemTools::CopyStatus SystemTools::CopyFileContentBlockwise(
   fin.close();
   fout.close();
 
+#if defined(__MVS__)
+  if (!SystemTools::CopyFileAttributes(source, destination)) {
+    return CopyStatus{ Status::POSIX_errno(), CopyStatus::DestPath };
+    }
+#endif
+
   if (!fout) {
     return CopyStatus{ Status::POSIX_errno(), CopyStatus::DestPath };
   }
@@ -2525,6 +2537,12 @@ SystemTools::CopyStatus SystemTools::CloneFileContent(
   close(in);
   close(out);
 
+#if defined(__MVS__)
+  if (!SystemTools::CopyFileAttributes(source, destination)) {
+    return CopyStatus{ Status::POSIX_errno(), CopyStatus::DestPath };
+    }
+#endif
+
   return status;
 #elif defined(__APPLE__) &&                                                   \
   defined(KWSYS_SYSTEMTOOLS_HAVE_MACOS_COPYFILE_CLONE)
@@ -2559,6 +2577,39 @@ SystemTools::CopyStatus SystemTools::CloneFileContent(
 #endif
 }
 
+#if defined(__MVS__)
+bool SystemTools::CopyFileAttributes(const std::string& source, const std::string& destination)
+{
+  struct stat sourceAttributes;
+  if ( stat(source.c_str(),&sourceAttributes) )
+    {
+    return false;
+    }
+
+  attrib_t destinationAttributes;
+  memset(&destinationAttributes, 0, sizeof(destinationAttributes));
+  destinationAttributes.att_filetagchg = 1;
+
+  // If source file is text, tag output file as IBM-819 text. Otherwise copy original tagging.
+  if ( sourceAttributes.st_tag.ft_txtflag == 1 )
+    {
+    destinationAttributes.att_filetag.ft_ccsid = 819;
+    destinationAttributes.att_filetag.ft_txtflag = 1;
+    }
+  else
+    {
+    destinationAttributes.att_filetag.ft_ccsid = sourceAttributes.st_tag.ft_ccsid;
+    destinationAttributes.att_filetag.ft_txtflag = 0;
+    }
+
+  if ( __chattr((char*)destination.c_str(), &destinationAttributes, sizeof(destinationAttributes)) )
+    {
+    return false;
+    }
+  return true;
+}
+#endif
+
 /**
  * Copy a file named by "source" to the file named by "destination".
  */
@@ -2609,6 +2660,7 @@ SystemTools::CopyStatus SystemTools::CopyFileAlways(
     if (!status.IsSuccess()) {
       status = SystemTools::CopyFileContentBlockwise(source, real_destination);
     }
+
     if (!status.IsSuccess()) {
       return status;
     }
diff --git a/Source/kwsys/SystemTools.hxx.in b/Source/kwsys/SystemTools.hxx.in
index 729928e6f3..800c8b206d 100644
--- a/Source/kwsys/SystemTools.hxx.in
+++ b/Source/kwsys/SystemTools.hxx.in
@@ -621,6 +621,13 @@ public:
    */
   static bool SameFile(const std::string& file1, const std::string& file2);
 
+#if defined(__MVS__)
+  /**
+   * Copy file attributes.
+   */
+  static bool CopyFileAttributes(const std::string& source, const std::string& destination);
+#endif
+
   /**
    * Copy a file.
    */
diff --git a/Utilities/cmcurl/include/curl/system.h b/Utilities/cmcurl/include/curl/system.h
index def7739242..eaf27020a4 100644
--- a/Utilities/cmcurl/include/curl/system.h
+++ b/Utilities/cmcurl/include/curl/system.h
@@ -237,10 +237,6 @@
 #  define CURL_PULL_SYS_SOCKET_H     1
 
 #elif defined(__MVS__)
-#  if defined(__IBMC__) || defined(__IBMCPP__)
-#    if defined(_ILP32)
-#    elif defined(_LP64)
-#    endif
 #    if defined(_LONG_LONG)
 #      define CURL_TYPEOF_CURL_OFF_T     long long
 #      define CURL_FORMAT_CURL_OFF_T     "lld"
@@ -263,7 +259,6 @@
 #    define CURL_TYPEOF_CURL_SOCKLEN_T socklen_t
 #    define CURL_PULL_SYS_TYPES_H      1
 #    define CURL_PULL_SYS_SOCKET_H     1
-#  endif
 
 #elif defined(__370__)
 #  if defined(__IBMC__) || defined(__IBMCPP__)
@@ -451,7 +446,6 @@
   typedef CURL_TYPEOF_CURL_SOCKLEN_T curl_socklen_t;
 #endif
 
-/* Data type definition of curl_off_t. */
 
 #ifdef CURL_TYPEOF_CURL_OFF_T
   typedef CURL_TYPEOF_CURL_OFF_T curl_off_t;
diff --git a/Utilities/cmcurl/lib/curl_setup.h b/Utilities/cmcurl/lib/curl_setup.h
index 9043d97e70..557a7d05ec 100644
--- a/Utilities/cmcurl/lib/curl_setup.h
+++ b/Utilities/cmcurl/lib/curl_setup.h
@@ -456,6 +456,10 @@
 #  endif
 #endif
 
+#ifdef __MVS__
+#define SIZEOF_CURL_OFF_T 8
+#endif
+
 #if (SIZEOF_CURL_OFF_T < 8)
 #error "too small curl_off_t"
 #else
diff --git a/Utilities/cmlibarchive/libarchive/filter_fork_posix.c b/Utilities/cmlibarchive/libarchive/filter_fork_posix.c
index 62085a7099..c72c926fc3 100644
--- a/Utilities/cmlibarchive/libarchive/filter_fork_posix.c
+++ b/Utilities/cmlibarchive/libarchive/filter_fork_posix.c
@@ -28,7 +28,7 @@
 
 /* This capability is only available on POSIX systems. */
 #if defined(HAVE_PIPE) && defined(HAVE_FCNTL) && \
-    (defined(HAVE_FORK) || defined(HAVE_VFORK) || defined(HAVE_POSIX_SPAWNP))
+    (defined(HAVE_FORK) || defined(HAVE_VFORK) || defined(HAVE_POSIX_SPAWNP)) || defined(__MVS__)
 
 __FBSDID("$FreeBSD: head/lib/libarchive/filter_fork.c 182958 2008-09-12 05:33:00Z kientzle $");
 
diff --git a/Utilities/cmlibuv/CMakeLists.txt b/Utilities/cmlibuv/CMakeLists.txt
index ad3d433822..3c6336d0b7 100644
--- a/Utilities/cmlibuv/CMakeLists.txt
+++ b/Utilities/cmlibuv/CMakeLists.txt
@@ -139,6 +139,33 @@ if(CMAKE_SYSTEM_NAME STREQUAL "AIX")
     )
 endif()
 
+if(CMAKE_SYSTEM_NAME STREQUAL "OS390")
+  list(APPEND uv_headers
+    include/uv/os390.h
+    )
+  enable_language(CXX)
+  list(APPEND uv_defines PATH_MAX=1024)
+  list(APPEND uv_defines _AE_BIMODAL)
+  list(APPEND uv_defines _ALL_SOURCE)
+  list(APPEND uv_defines _ISOC99_SOURCE)
+  list(APPEND uv_defines _LARGE_TIME_API)
+  list(APPEND uv_defines _OPEN_MSGQ_EXT)
+  list(APPEND uv_defines _OPEN_SYS_FILE_EXT)
+  list(APPEND uv_defines _OPEN_SYS_IF_EXT)
+  list(APPEND uv_defines _OPEN_SYS_SOCK_EXT3)
+  list(APPEND uv_defines _OPEN_SYS_SOCK_IPV6)
+  list(APPEND uv_defines _UNIX03_SOURCE)
+  list(APPEND uv_defines _UNIX03_THREADS)
+  list(APPEND uv_defines _UNIX03_WITHDRAWN)
+  list(APPEND uv_defines _XOPEN_SOURCE=600)
+  list(APPEND uv_defines _XOPEN_SOURCE_EXTENDED)
+  list(APPEND uv_sources
+       src/unix/pthread-fixes.c
+       src/unix/os390.c
+       src/unix/os390-syscalls.c
+       src/unix/os390-proctitle.c)
+endif()
+
 if(CMAKE_SYSTEM_NAME STREQUAL "OS400")
   list(APPEND uv_headers
     include/uv/posix.h
diff --git a/Utilities/cmlibuv/include/uv/posix.h b/Utilities/cmlibuv/include/uv/posix.h
index 9a96634db0..98248bff05 100644
--- a/Utilities/cmlibuv/include/uv/posix.h
+++ b/Utilities/cmlibuv/include/uv/posix.h
@@ -22,6 +22,10 @@
 #ifndef UV_POSIX_H
 #define UV_POSIX_H
 
+#ifdef __MVS__
+#define UV_PLATFORM_SEM_T long
+#endif
+
 #define UV_PLATFORM_LOOP_FIELDS                                               \
   struct pollfd* poll_fds;                                                    \
   size_t poll_fds_used;                                                       \
diff --git a/Utilities/cmlibuv/src/unix/core.c b/Utilities/cmlibuv/src/unix/core.c
index d0b0e0069b..918b1ed05c 100644
--- a/Utilities/cmlibuv/src/unix/core.c
+++ b/Utilities/cmlibuv/src/unix/core.c
@@ -88,6 +88,7 @@ extern char** environ;
 
 #if defined(__MVS__)
 # include <sys/ioctl.h>
+# include <stdbool.h>
 # include "zos-sys-info.h"
 #endif
 
diff --git a/Utilities/cmlibuv/src/unix/os390-syscalls.h b/Utilities/cmlibuv/src/unix/os390-syscalls.h
index 9f504171d8..d467af9c45 100644
--- a/Utilities/cmlibuv/src/unix/os390-syscalls.h
+++ b/Utilities/cmlibuv/src/unix/os390-syscalls.h
@@ -28,7 +28,6 @@
 #include <dirent.h>
 #include <poll.h>
 #include <pthread.h>
-#include "zos-base.h"
 
 #define EPOLL_CTL_ADD             1
 #define EPOLL_CTL_DEL             2
@@ -58,6 +57,7 @@ int epoll_wait(uv__os390_epoll* ep, struct epoll_event *events, int maxevents, i
 int epoll_file_close(int fd);
 
 /* utility functions */
+int nanosleep(const struct timespec* req, struct timespec* rem);
 int scandir(const char* maindir, struct dirent*** namelist,
             int (*filter)(const struct dirent *),
             int (*compar)(const struct dirent **,
diff --git a/Utilities/cmzlib/zconf.h b/Utilities/cmzlib/zconf.h
index 1790072cf7..afbe26230f 100644
--- a/Utilities/cmzlib/zconf.h
+++ b/Utilities/cmzlib/zconf.h
@@ -516,21 +516,4 @@ typedef uLong FAR uLongf;
 #  endif
 #endif
 
-/* MVS linker does not support external names larger than 8 bytes */
-#if defined(__MVS__)
-  #pragma map(deflateInit_,"DEIN")
-  #pragma map(deflateInit2_,"DEIN2")
-  #pragma map(deflateEnd,"DEEND")
-  #pragma map(deflateBound,"DEBND")
-  #pragma map(inflateInit_,"ININ")
-  #pragma map(inflateInit2_,"ININ2")
-  #pragma map(inflateEnd,"INEND")
-  #pragma map(inflateSync,"INSY")
-  #pragma map(inflateSetDictionary,"INSEDI")
-  #pragma map(compressBound,"CMBND")
-  #pragma map(inflate_table,"INTABL")
-  #pragma map(inflate_fast,"INFA")
-  #pragma map(inflate_copyright,"INCOPY")
-#endif
-
 #endif /* ZCONF_H */
diff --git a/Utilities/cmzstd/lib/common/debug.h b/Utilities/cmzstd/lib/common/debug.h
index 3b2a320a18..7b8b7fd4f0 100644
--- a/Utilities/cmzstd/lib/common/debug.h
+++ b/Utilities/cmzstd/lib/common/debug.h
@@ -70,7 +70,10 @@ extern "C" {
 #  define ZSTD_DEPS_NEED_ASSERT
 #  include "zstd_deps.h"
 #else
-#  ifndef assert   /* assert may be already defined, due to prior #include <assert.h> */
+#  if !defined(assert) /* assert may be already defined, due to prior #include <assert.h> */
+#  ifdef __MVS__ 
+#    define __assert 1 /* Indicate to the headers that we have our own assert */
+#  endif
 #    define assert(condition) ((void)0)   /* disable assert (default) */
 #  endif
 #endif
diff --git a/Utilities/cmzstd/lib/common/zstd_trace.h b/Utilities/cmzstd/lib/common/zstd_trace.h
index 485cadf7c4..c176a3740f 100644
--- a/Utilities/cmzstd/lib/common/zstd_trace.h
+++ b/Utilities/cmzstd/lib/common/zstd_trace.h
@@ -20,7 +20,7 @@ extern "C" {
 /* weak symbol support */
 #if !defined(ZSTD_HAVE_WEAK_SYMBOLS) && defined(__GNUC__) && \
     !defined(__APPLE__) && !defined(_WIN32) && !defined(__MINGW32__) && \
-    !defined(__CYGWIN__) && !defined(_AIX)
+    !defined(__CYGWIN__) && !defined(_AIX) && !defined(__MVS__)
 #  define ZSTD_HAVE_WEAK_SYMBOLS 1
 #else
 #  define ZSTD_HAVE_WEAK_SYMBOLS 0
diff --git a/bootstrap b/bootstrap
index cc604256d8..8f1b8ff0c4 100755
--- a/bootstrap
+++ b/bootstrap
@@ -172,6 +172,13 @@ else
   cmake_system_linux=false
 fi
 
+# Determine whether this is z/OS
+if echo "${cmake_system}" | grep OS/390 >/dev/null 2>&1; then
+  cmake_system_zos=true
+else
+  cmake_system_zos=false
+ fi
+
 # Determine whether this is a PA-RISC machine
 # This only works for Linux or HP-UX, not other PA-RISC OSs (BSD maybe?). Also
 # may falsely detect parisc on HP-UX m68k
@@ -647,6 +654,14 @@ else
     "
 fi
 
+if ${cmake_system_zos}; then
+  LIBUV_C_SOURCES="${LIBUV_C_SOURCES}\
+    src/unix/os390-syscalls.c \
+    src/unix/pthread-fixes.c \
+    src/unix/os390-proctitle.c \
+  "
+fi
+
 # Display CMake bootstrap usage
 cmake_usage()
 {
@@ -986,6 +1001,7 @@ while test $# != 0; do
   CFLAGS=*) CFLAGS=`cmake_arg "$1"` ;;
   CXXFLAGS=*) CXXFLAGS=`cmake_arg "$1"` ;;
   LDFLAGS=*) LDFLAGS=`cmake_arg "$1"` ;;
+  LIBS=*) LIBS=`cmake_arg "$1"` ;;
   --) shift; break ;;
   *) die "Unknown option: $1" ;;
   esac
@@ -1151,6 +1167,7 @@ case "${cmake_system}" in
   *Darwin*) cmake_toolchains='Clang GNU' ;;
   *Linux*) cmake_toolchains='GNU Clang XL PGI PathScale' ;;
   *MINGW*) cmake_toolchains='GNU' ;;
+  *OS/390*) cmake_toolchains='Clang z/OS' ;;
 esac
 
 # Toolchain compiler name table.
@@ -1733,7 +1750,7 @@ if test "${cmake_bootstrap_generator}" = "Ninja"; then
   fi
 fi
 
-libs=""
+libs="${LIBS}"
 
 uv_c_flags=""
 if ${cmake_system_mingw}; then
@@ -1771,6 +1788,13 @@ else
       uv_c_flags="${uv_c_flags} -D_XOPEN_SOURCE=700"
       libs="${libs} -lsocket"
       ;;
+    *OS/390*)
+      uv_c_flags="${uv_c_flags} -D_XOPEN_SOURCE=600 -DPATH_MAX=1024 -D_AE_BIMODAL -D_ALL_SOURCE \
+                  -D_ISOC99_SOURCE -D_LARGE_TIME_API -D_OPEN_MSGQ_EXT -D_OPEN_SYS_FILE_EXT -D_OPEN_SYS_IF_EXT \
+                  -D_OPEN_SYS_SOCK_EXT3 -D_OPEN_SYS_SOCK_IPV6 -D_UNIX03_SOURCE -D_UNIX03_THREADS -D_UNIX03_WITHDRAWN \
+                  -D_XOPEN_SOURCE_EXTENDED"
+      libs="${libs}"
+      ;;
   esac
 fi
 if test "x${bootstrap_system_libuv}" = "x"; then
@@ -2075,6 +2099,7 @@ export MAKE
 export CFLAGS
 export CXXFLAGS
 export LDFLAGS
+export LIBS
 
 # Run bootstrap CMake to configure real CMake
 cmake_options="-DCMAKE_BOOTSTRAP=1"
diff --git a/Source/cmSystemTools.cxx b/Source/cmSystemTools.cxx
index 1fb0079103..ae00fad071 100644
--- a/Source/cmSystemTools.cxx
+++ b/Source/cmSystemTools.cxx
@@ -1,7 +1,7 @@
 /* Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
    file Copyright.txt or https://cmake.org/licensing for details.  */
 
-#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__)
+#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__) && !defined(__MVS__)
 // POSIX APIs are needed
 // NOLINTNEXTLINE(bugprone-reserved-identifier)
 #  define _POSIX_C_SOURCE 200809L
@@ -3765,6 +3765,14 @@ cm::string_view cmSystemTools::GetSystemName()
     if (systemName.find("MSYS") != cm::string_view::npos) {
       systemName = "MSYS";
     }
+
+    // fix for OS/390, remove the /
+    static const cmsys::RegularExpression os390Regex("OS.390");
+    cmsys::RegularExpressionMatch match390;
+    if (os390Regex.find(uts_name.sysname, match390)) {
+      systemName = "OS390";
+    }
+
     return systemName;
   }
   return "";
