diff --git a/Utilities/cmcurl/include/curl/system.h b/Utilities/cmcurl/include/curl/system.h
index def7739242..eaf27020a4 100644
--- a/Utilities/cmcurl/include/curl/system.h
+++ b/Utilities/cmcurl/include/curl/system.h
@@ -237,10 +237,6 @@
 #  define CURL_PULL_SYS_SOCKET_H     1
 
 #elif defined(__MVS__)
-#  if defined(__IBMC__) || defined(__IBMCPP__)
-#    if defined(_ILP32)
-#    elif defined(_LP64)
-#    endif
 #    if defined(_LONG_LONG)
 #      define CURL_TYPEOF_CURL_OFF_T     long long
 #      define CURL_FORMAT_CURL_OFF_T     "lld"
@@ -263,7 +259,6 @@
 #    define CURL_TYPEOF_CURL_SOCKLEN_T socklen_t
 #    define CURL_PULL_SYS_TYPES_H      1
 #    define CURL_PULL_SYS_SOCKET_H     1
-#  endif
 
 #elif defined(__370__)
 #  if defined(__IBMC__) || defined(__IBMCPP__)
@@ -451,7 +446,6 @@
   typedef CURL_TYPEOF_CURL_SOCKLEN_T curl_socklen_t;
 #endif
 
-/* Data type definition of curl_off_t. */
 
 #ifdef CURL_TYPEOF_CURL_OFF_T
   typedef CURL_TYPEOF_CURL_OFF_T curl_off_t;
