diff --git a/Source/kwsys/SystemTools.hxx.in b/Source/kwsys/SystemTools.hxx.in
index 729928e6f3..800c8b206d 100644
--- a/Source/kwsys/SystemTools.hxx.in
+++ b/Source/kwsys/SystemTools.hxx.in
@@ -621,6 +621,13 @@ public:
    */
   static bool SameFile(const std::string& file1, const std::string& file2);
 
+#if defined(__MVS__)
+  /**
+   * Copy file attributes.
+   */
+  static bool CopyFileAttributes(const std::string& source, const std::string& destination);
+#endif
+
   /**
    * Copy a file.
    */
