diff --git a/Modules/Compiler/IBMClang-CXX.cmake b/Modules/Compiler/IBMClang-CXX.cmake
index 5431b17e20..16ba6e65f7 100644
--- a/Modules/Compiler/IBMClang-CXX.cmake
+++ b/Modules/Compiler/IBMClang-CXX.cmake
@@ -26,14 +26,19 @@ set(CMAKE_CXX14_STANDARD__HAS_FULL_SUPPORT ON)
 set(CMAKE_CXX14_STANDARD_COMPILE_OPTION "-std=c++14")
 set(CMAKE_CXX14_EXTENSION_COMPILE_OPTION "-std=gnu++14")
 
-if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 17.1.0)
+if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 17.1.0 OR
+    (CMAKE_SYSTEM_NAME STREQUAL "OS390" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 1.1.0))
   set(CMAKE_CXX17_STANDARD_COMPILE_OPTION  "-std=c++17")
   set(CMAKE_CXX17_EXTENSION_COMPILE_OPTION "-std=gnu++17")
   set(CMAKE_CXX20_STANDARD_COMPILE_OPTION  "-std=c++20")
   set(CMAKE_CXX20_EXTENSION_COMPILE_OPTION "-std=gnu++20")
 endif()
 
-__compiler_check_default_language_standard(CXX 17.1.0 17)
+if (CMAKE_SYSTEM_NAME STREQUAL "OS390")
+  __compiler_check_default_language_standard(CXX 1.1.0 14)
+else()
+  __compiler_check_default_language_standard(CXX 17.1.0 17)
+endif()
 
 set(CMAKE_CXX_COMPILE_OBJECT
   "<CMAKE_CXX_COMPILER> -x c++ <DEFINES> <INCLUDES> <FLAGS> -o <OBJECT> -c <SOURCE>")
