diff --git a/Modules/Compiler/zOS-C.cmake b/Modules/Compiler/zOS-C.cmake
new file mode 100644
index 0000000000..73928346d2
--- /dev/null
+++ b/Modules/Compiler/zOS-C.cmake
@@ -0,0 +1,7 @@
+include(Compiler/zOS)
+__compiler_zos(C)
+set(CMAKE_C_FLAGS_RELEASE_INIT "${CMAKE_C_FLAGS_RELEASE_INIT} -DNDEBUG")
+set(CMAKE_C_FLAGS_MINSIZEREL_INIT "${CMAKE_C_FLAGS_MINSIZEREL_INIT} -DNDEBUG")
+
+# Raise severity of include file checks from warning to error for C.
+set(CMAKE_PLATFORM_CHECK_INCLUDE_C_FLAGS "-qseverity=e=ccn3296")
