diff --git a/Modules/Compiler/zOS-CXX.cmake b/Modules/Compiler/zOS-CXX.cmake
new file mode 100644
index 0000000000..de82177f25
--- /dev/null
+++ b/Modules/Compiler/zOS-CXX.cmake
@@ -0,0 +1,39 @@
+include(Compiler/zOS)
+__compiler_zos(CXX)
+set(CMAKE_CXX_FLAGS_RELEASE_INIT "${CMAKE_CXX_FLAGS_RELEASE_INIT} -DNDEBUG")
+set(CMAKE_CXX_FLAGS_MINSIZEREL_INIT "${CMAKE_CXX_FLAGS_MINSIZEREL_INIT} -DNDEBUG")
+
+if(NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 2.3.1)
+  set(CMAKE_CXX98_STANDARD__HAS_FULL_SUPPORT ON)
+  set(CMAKE_CXX98_STANDARD_COMPILE_OPTION "-std=c++98")
+  set(CMAKE_CXX98_EXTENSION_COMPILE_OPTION "-std=gnu++98")
+
+  set(CMAKE_CXX11_STANDARD__HAS_FULL_SUPPORT ON)
+  set(CMAKE_CXX11_STANDARD_COMPILE_OPTION "-std=c++11")
+  set(CMAKE_CXX11_EXTENSION_COMPILE_OPTION "-std=gnu++11")
+
+  set(CMAKE_CXX14_STANDARD__HAS_FULL_SUPPORT ON)
+  set(CMAKE_CXX14_STANDARD_COMPILE_OPTION "-std=c++14")
+  set(CMAKE_CXX14_EXTENSION_COMPILE_OPTION "-std=gnu++14")
+endif()
+
+macro(cmake_record_cxx_compile_features)
+  list(APPEND CMAKE_CXX_COMPILE_FEATURES
+    cxx_std_98
+    cxx_std_11
+    cxx_std_14
+    )
+  _record_compiler_features(CXX "" CMAKE_CXX_COMPILE_FEATURES)
+endmacro()
+
+if(NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 2.3.1)
+  if (NOT CMAKE_CXX_COMPILER_FORCED)
+    if (NOT CMAKE_CXX_STANDARD_COMPUTED_DEFAULT)
+      message(FATAL_ERROR "CMAKE_CXX_STANDARD_COMPUTED_DEFAULT should be set for ${CMAKE_CXX_COMPILER_ID} (${CMAKE_CXX_COMPILER}) version ${CMAKE_CXX_COMPILER_VERSION}")
+    endif()
+    set(CMAKE_CXX_STANDARD_DEFAULT ${CMAKE_CXX_STANDARD_COMPUTED_DEFAULT})
+  elseif(NOT DEFINED CMAKE_CXX_STANDARD_DEFAULT)
+    # Compiler id was forced so just guess the default standard level.
+    set(CMAKE_CXX_STANDARD_DEFAULT 98)
+  endif()
+endif()
